app/controllers/admin/site_settings_controller.rb

```diff
@@ -37,127 +40,124 @@ def update
 
     id = new_setting_name if new_setting_name
 
-    raise_access_hidden_setting(id)
-
-    case SiteSetting.type_supervisor.get_type(id)
-    when :integer
-      value = value.tr("^-0-9", "")
-    when :file_size_restriction
-      value = value.tr("^0-9", "").to_i
-    when :uploaded_image_list
-      value = value.blank? ? "" : Upload.get_from_urls(value.split("|")).to_a
-    end
-
-    value = Upload.get_from_url(value) || "" if SiteSetting.type_supervisor.get_type(id) == :upload
-
-    update_existing_users = params[:update_existing_user].present?
     previous_value = value_or_default(SiteSetting.get(id)) if update_existing_users
 
-    SiteSetting.set_and_log(id, value, current_user)
-
-    if update_existing_users
-      new_value = value_or_default(value)
-
-      if (user_option = user_options[id.to_sym]).present?
-        if user_option == "text_size_key"
-          previous_value = UserOption.text_sizes[previous_value.to_sym]
-          new_value = UserOption.text_sizes[new_value.to_sym]
-        elsif user_option == "title_count_mode_key"
-          previous_value = UserOption.title_count_modes[previous_value.to_sym]
-          new_value = UserOption.title_count_modes[new_value.to_sym]
-        end
+    with_service(UpdateSiteSetting, setting_name: id, new_value: value, current_user:) do
+      on_success do
+        value = result.new_value
```

Comment:
Okay this was a bit of a jump scare :ghost: Let's move all this logic to an "Action" class which are basically a lightweight version of services without `Service::Base`, more like our "traditional" services. You can see an example at https://github.com/discourse/discourse/blob/e92323c81e0531797aedb1725f233b3c29c50164/plugins/chat/app/services/chat/action/calculate_memberships_for_removal.rb . It should have a single `self.call` method.

Then we can do `SiteSettingUpdateExistingUsers.call(setting: id, new_value: value, previous_value: previous_value)` and remove all this ugliness from the controller, so it can also be tested in isolation.

This whole "update existing users" thing also ties in to what I am working on for requiring confirmation for certain dangerous site setting changes, so I may want to move it around even more in future. Having it in a distinct class instead of jammed into the controller will help that a lot.